File: backend/ITSAssetTracker.API/Controllers/InventoryController.cs
using ITSAssetTracker.API.Data;
using ITSAssetTracker.API.Middleware;
using ITSAssetTracker.API.Models;
using ITSAssetTracker.API.Models.DTOs;
using ITSAssetTracker.API.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace ITSAssetTracker.API.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class InventoryController : ControllerBase
{
    private readonly AssetTrackerDbContext _context;
    private readonly IEmailService _emailService;
    private readonly IConfiguration _configuration;
    private readonly ILogger<InventoryController> _logger;

    public InventoryController(
        AssetTrackerDbContext context,
        IEmailService emailService,
        IConfiguration configuration,
        ILogger<InventoryController> logger)
    {
        _context = context;
        _emailService = emailService;
        _configuration = configuration;
        _logger = logger;
    }

    [HttpGet]
    public async Task<ActionResult<IEnumerable<InventoryDto>>> GetAll(
        [FromQuery] string? search,
        [FromQuery] string? assetType,
        [FromQuery] string? category,
        [FromQuery] string? hardwareType, // Legacy parameter
        [FromQuery] bool? needsReorder,
        [FromQuery] string? sortBy,
        [FromQuery] bool sortDesc = false)
    {
        var query = _context.Inventories.AsQueryable();

        // Apply filters
        if (!string.IsNullOrEmpty(search))
        {
            query = query.Where(i =>
                i.ItemNumber.Contains(search) ||
                (i.Description != null && i.Description.Contains(search)) ||
                (i.HardwareDescription != null && i.HardwareDescription.Contains(search)));
        }

        // New asset type filter
        if (!string.IsNullOrEmpty(assetType))
        {
            query = query.Where(i => i.AssetType == assetType);
        }

        // New category filter
        if (!string.IsNullOrEmpty(category))
        {
            query = query.Where(i => i.Category == category || i.HardwareType == category);
        }

        // Legacy hardware type filter (for backward compatibility)
        if (!string.IsNullOrEmpty(hardwareType))
        {
            query = query.Where(i => i.HardwareType == hardwareType || i.Category == hardwareType);
        }

        if (needsReorder == true)
        {
            query = query.Where(i => i.CurrentQuantity <= i.MinimumThreshold);
        }

        // Apply sorting
        query = sortBy?.ToLower() switch
        {
            "itemnumber" => sortDesc ? query.OrderByDescending(i => i.ItemNumber) : query.OrderBy(i => i.ItemNumber),
            "description" => sortDesc ? query.OrderByDescending(i => i.Description ?? i.HardwareDescription) : query.OrderBy(i => i.Description ?? i.HardwareDescription),
            "type" => sortDesc ? query.OrderByDescending(i => i.Category ?? i.HardwareType) : query.OrderBy(i => i.Category ?? i.HardwareType),
            "assettype" => sortDesc ? query.OrderByDescending(i => i.AssetType) : query.OrderBy(i => i.AssetType),
            "quantity" => sortDesc ? query.OrderByDescending(i => i.CurrentQuantity) : query.OrderBy(i => i.CurrentQuantity),
            "cost" => sortDesc ? query.OrderByDescending(i => i.Cost) : query.OrderBy(i => i.Cost),
            "modified" => sortDesc ? query.OrderByDescending(i => i.LastModifiedDate) : query.OrderBy(i => i.LastModifiedDate),
            _ => query.OrderBy(i => i.ItemNumber)
        };

        var items = await query.ToListAsync();

        var dtos = items.Select(i => new InventoryDto
        {
            Id = i.Id,
            ItemNumber = i.ItemNumber,
            AssetType = i.AssetType,
            Description = i.Description,
            Category = i.Category,
            Cost = i.Cost,
            MinimumThreshold = i.MinimumThreshold,
            ReorderAmount = i.ReorderAmount,
            CurrentQuantity = i.CurrentQuantity,
            LastModifiedBy = i.LastModifiedBy,
            LastModifiedDate = i.LastModifiedDate,
            // Legacy fields for backward compatibility
            HardwareDescription = i.HardwareDescription ?? i.Description,
            HardwareType = i.HardwareType ?? i.Category
        }).ToList();

        return Ok(dtos);
    }

    [HttpGet("{id}")]
    public async Task<ActionResult<InventoryDto>> GetById(int id)
    {
        var item = await _context.Inventories.FindAsync(id);

        if (item == null)
        {
            return NotFound();
        }

        var dto = new InventoryDto
        {
            Id = item.Id,
            ItemNumber = item.ItemNumber,
            AssetType = item.AssetType,
            Description = item.Description,
            Category = item.Category,
            Cost = item.Cost,
            MinimumThreshold = item.MinimumThreshold,
            ReorderAmount = item.ReorderAmount,
            CurrentQuantity = item.CurrentQuantity,
            LastModifiedBy = item.LastModifiedBy,
            LastModifiedDate = item.LastModifiedDate,
            HardwareDescription = item.HardwareDescription ?? item.Description,
            HardwareType = item.HardwareType ?? item.Category
        };

        return Ok(dto);
    }

    [HttpGet("{id}/audit-history")]
    public async Task<ActionResult<IEnumerable<AuditHistory>>> GetAuditHistory(int id)
    {
        var history = await _context.AuditHistories
            .Where(a => a.ItemId == id)
            .OrderByDescending(a => a.ChangeDate)
            .ToListAsync();

        return Ok(history);
    }

    [HttpGet("types")]
    public async Task<ActionResult<IEnumerable<string>>> GetHardwareTypes()
    {
        // Return all unique categories (supports both legacy HardwareType and new Category)
        var types = await _context.Inventories
            .Select(i => i.Category ?? i.HardwareType)
            .Where(t => t != null)
            .Distinct()
            .OrderBy(t => t)
            .ToListAsync();

        return Ok(types);
    }

    [HttpGet("asset-types")]
    public async Task<ActionResult<IEnumerable<string>>> GetAssetTypes()
    {
        var assetTypes = await _context.Inventories
            .Select(i => i.AssetType)
            .Distinct()
            .OrderBy(t => t)
            .ToListAsync();

        return Ok(assetTypes);
    }

    [HttpGet("low-stock-count")]
    public async Task<ActionResult<int>> GetLowStockCount()
    {
        var count = await _context.Inventories
            .CountAsync(i => i.CurrentQuantity <= i.MinimumThreshold);

        return Ok(count);
    }

    [HttpPost("update-quantity")]
    [ADGroupAuthorization("IT_ServiceDesk")]
    public async Task<ActionResult> UpdateQuantity([FromBody] UpdateInventoryDto dto)
    {
        var username = User.Identity?.Name ?? "Unknown";

        var item = await _context.Inventories
            .FirstOrDefaultAsync(i => i.ItemNumber == dto.ItemNumber);

        if (item == null)
        {
            return NotFound($"Item {dto.ItemNumber} not found");
        }

        var previousQuantity = item.CurrentQuantity;
        var newQuantity = previousQuantity + dto.QuantityChange;

        if (newQuantity < 0)
        {
            return BadRequest("Quantity cannot be negative");
        }

        item.CurrentQuantity = newQuantity;
        item.LastModifiedBy = username;
        item.LastModifiedDate = DateTime.UtcNow;

        // Create audit entry
        var auditEntry = new AuditHistory
        {
            ItemId = item.Id,
            PreviousQuantity = previousQuantity,
            NewQuantity = newQuantity,
            ChangedBy = username,
            ChangeDate = DateTime.UtcNow,
            ServiceNowTicketUrl = dto.ServiceNowTicketUrl
        };

        _context.AuditHistories.Add(auditEntry);
        await _context.SaveChangesAsync();

        // Send alert if item is now below threshold
        if (newQuantity <= item.MinimumThreshold && previousQuantity > item.MinimumThreshold)
        {
            var appBaseUrl = _configuration["AppSettings:BaseUrl"] ?? "http://localhost:3000";
            await _emailService.SendLowStockAlert(item, appBaseUrl);
        }

        _logger.LogInformation("Inventory updated: {ItemNumber} from {Old} to {New} by {User}",
            dto.ItemNumber, previousQuantity, newQuantity, username);

        return Ok(new { previousQuantity, newQuantity, itemNumber = item.ItemNumber });
    }

    [HttpGet("dashboard-stats")]
    public async Task<ActionResult> GetDashboardStats()
    {
        var totalItems = await _context.Inventories.CountAsync();
        var lowStockCount = await _context.Inventories
            .CountAsync(i => i.CurrentQuantity <= i.MinimumThreshold);
        var totalValue = await _context.Inventories
            .SumAsync(i => i.Cost * i.CurrentQuantity);
        var recentChanges = await _context.AuditHistories
            .OrderByDescending(a => a.ChangeDate)
            .Take(10)
            .Include(a => a.Item)
            .ToListAsync();

        return Ok(new
        {
            totalItems,
            lowStockCount,
            totalValue,
            recentChanges
        });
    }
}

File: backend-express/src/controllers/inventory.controller.ts
import { Request, Response } from 'express';
import prisma from '../config/prisma';
import { emailService } from '../services/email.service';
import { config } from '../config';
import { validateInventoryUpdate } from '../validation/inventory.schema';

export class InventoryController {
  /**
   * GET /api/inventory
   * Get all inventory items with filtering and sorting
   */
  async getAll(req: Request, res: Response) {
    try {
      const {
        search,
        assetType,
        category,
        hardwareType,
        needsReorder,
        sortBy = 'itemNumber',
        sortDesc = 'false',
      } = req.query;

      // Build where clause
      const where: any = {};

      if (search) {
        where.OR = [
          { itemNumber: { contains: search as string, mode: 'insensitive' } },
          { description: { contains: search as string, mode: 'insensitive' } },
          { hardwareDescription: { contains: search as string, mode: 'insensitive' } },
        ];
      }

      if (assetType) {
        where.assetType = assetType as string;
      }

      if (category) {
        where.category = category as string;
      }

      if (hardwareType) {
        where.hardwareType = hardwareType as string;
      }

      // Get all items
      let items = await prisma.inventory.findMany({
        where,
        orderBy: {
          [sortBy as string]: sortDesc === 'true' ? 'desc' : 'asc',
        },
      });

      // Filter by needsReorder if specified
      if (needsReorder === 'true') {
        items = items.filter(item => item.currentQuantity < item.minimumThreshold);
      } else if (needsReorder === 'false') {
        items = items.filter(item => item.currentQuantity >= item.minimumThreshold);
      }

      // Add needsReorder property to each item
      const itemsWithReorderFlag = items.map(item => ({
        ...item,
        needsReorder: item.currentQuantity < item.minimumThreshold,
      }));

      res.json(itemsWithReorderFlag);
    } catch (error: any) {
      console.error('Error fetching inventory:', error);
      res.status(500).json({ error: 'Failed to fetch inventory items' });
    }
  }

  /**
   * GET /api/inventory/:id
   * Get single inventory item by ID
   */
  async getById(req: Request, res: Response) {
    try {
      const { id } = req.params;

      const item = await prisma.inventory.findUnique({
        where: { id: parseInt(id, 10) },
      });

      if (!item) {
        return res.status(404).json({ error: 'Inventory item not found' });
      }

      res.json({
        ...item,
        needsReorder: item.currentQuantity < item.minimumThreshold,
      });
    } catch (error: any) {
      console.error('Error fetching inventory item:', error);
      res.status(500).json({ error: 'Failed to fetch inventory item' });
    }
  }

  /**
   * GET /api/inventory/audit-history
   * Get all audit history with optional filtering
   */
  async getAllAuditHistory(req: Request, res: Response) {
    try {
      const { limit, search } = req.query;

      const where: any = {};

      if (search) {
        where.item = {
          OR: [
            { itemNumber: { contains: search as string, mode: 'insensitive' } },
            { description: { contains: search as string, mode: 'insensitive' } },
          ],
        };
      }

      const auditHistory = await prisma.auditHistory.findMany({
        where,
        include: {
          item: true,
        },
        orderBy: {
          changeDate: 'desc',
        },
        take: limit ? parseInt(limit as string, 10) : undefined,
      });

      res.json(auditHistory);
    } catch (error: any) {
      console.error('Error fetching audit history:', error);
      res.status(500).json({ error: 'Failed to fetch audit history' });
    }
  }

  /**
   * GET /api/inventory/:id/audit-history
   * Get audit history for specific item
   */
  async getAuditHistoryByItemId(req: Request, res: Response) {
    try {
      const { id } = req.params;

      const auditHistory = await prisma.auditHistory.findMany({
        where: { itemId: parseInt(id, 10) },
        orderBy: { changeDate: 'desc' },
      });

      res.json(auditHistory);
    } catch (error: any) {
      console.error('Error fetching item audit history:', error);
      res.status(500).json({ error: 'Failed to fetch item audit history' });
    }
  }

  /**
   * GET /api/inventory/types
   * Get all unique hardware types/categories
   */
  async getTypes(req: Request, res: Response) {
    try {
      const items = await prisma.inventory.findMany({
        select: {
          hardwareType: true,
          category: true,
        },
      });

      // Combine and deduplicate types
      const typesSet = new Set<string>();
      items.forEach(item => {
        if (item.hardwareType) typesSet.add(item.hardwareType);
        if (item.category) typesSet.add(item.category);
      });

      const types = Array.from(typesSet).filter(Boolean).sort();

      res.json(types);
    } catch (error: any) {
      console.error('Error fetching inventory types:', error);
      res.status(500).json({ error: 'Failed to fetch inventory types' });
    }
  }

  /**
   * GET /api/inventory/asset-types
   * Get all asset types (Hardware/Software)
   */
  async getAssetTypes(req: Request, res: Response) {
    try {
      const items = await prisma.inventory.findMany({
        select: {
          assetType: true,
        },
        distinct: ['assetType'],
      });

      const assetTypes = items.map(item => item.assetType).filter(Boolean).sort();

      res.json(assetTypes);
    } catch (error: any) {
      console.error('Error fetching asset types:', error);
      res.status(500).json({ error: 'Failed to fetch asset types' });
    }
  }

  /**
   * GET /api/inventory/low-stock-count
   * Get count of items below threshold
   */
  async getLowStockCount(req: Request, res: Response) {
    try {
      const items = await prisma.inventory.findMany();
      const lowStockCount = items.filter(item => item.currentQuantity < item.minimumThreshold).length;

      res.json({ count: lowStockCount });
    } catch (error: any) {
      console.error('Error fetching low stock count:', error);
      res.status(500).json({ error: 'Failed to fetch low stock count' });
    }
  }

  /**
   * GET /api/inventory/dashboard-stats
   * Get dashboard statistics
   */
  async getDashboardStats(req: Request, res: Response) {
    try {
      const items = await prisma.inventory.findMany();

      const totalItems = items.reduce((sum, item) => sum + item.currentQuantity, 0);
      const lowStockCount = items.filter(item => item.currentQuantity < item.minimumThreshold).length;
      const totalValue = items.reduce((sum, item) => sum + item.cost * item.currentQuantity, 0);

      // Get recent changes (last 10 audit entries)
      const recentChanges = await prisma.auditHistory.findMany({
        take: 10,
        orderBy: { changeDate: 'desc' },
        include: { item: true },
      });

      res.json({
        totalItems,
        lowStockCount,
        totalValue,
        recentChanges,
      });
    } catch (error: any) {
      console.error('Error fetching dashboard stats:', error);
      res.status(500).json({ error: 'Failed to fetch dashboard stats' });
    }
  }

  /**
   * POST /api/inventory/update-quantity
   * Update inventory quantity
   */
  async updateQuantity(req: Request, res: Response) {
    try {
      const { itemNumber, quantityChange, serviceNowTicketUrl, assignedToUser } = req.body;

      if (!itemNumber || quantityChange === undefined) {
        return res.status(400).json({ error: 'ItemNumber and QuantityChange are required' });
      }

      const item = await prisma.inventory.findUnique({
        where: { itemNumber },
      });

      if (!item) {
        return res.status(404).json({ error: 'Inventory item not found' });
      }

      const previousQuantity = item.currentQuantity;
      const newQuantity = previousQuantity + quantityChange;

      if (newQuantity < 0) {
        return res.status(400).json({ error: 'Resulting quantity cannot be negative' });
      }

      // Update inventory
      const updatedItem = await prisma.inventory.update({
        where: { id: item.id },
        data: {
          currentQuantity: newQuantity,
          lastModifiedBy: req.user?.username || 'System',
          lastModifiedDate: new Date(),
        },
      });

      // Create audit entry
      await prisma.auditHistory.create({
        data: {
          itemId: item.id,
          itemNumber: item.itemNumber,
          itemDescription: item.description || item.hardwareDescription || 'No description',
          previousQuantity,
          newQuantity,
          changedBy: req.user?.username || 'System',
          changeDate: new Date(),
          serviceNowTicketUrl: serviceNowTicketUrl || null,
        },
      });

      // Send low stock alert if needed
      if (newQuantity < item.minimumThreshold && previousQuantity >= item.minimumThreshold) {
        try {
          await emailService.sendLowStockAlert(updatedItem, config.baseUrl);
        } catch (emailError) {
          console.error('Error sending low stock alert:', emailError);
          // Don't fail the request if email fails
        }
      }

      res.json({
        ...updatedItem,
        needsReorder: updatedItem.currentQuantity < updatedItem.minimumThreshold,
      });
    } catch (error: any) {
      console.error('Error updating quantity:', error);
      res.status(500).json({ error: 'Failed to update inventory quantity' });
    }
  }

  /**
   * PUT /api/inventory/:id
   * Update inventory item
   */
  async updateItem(req: Request, res: Response) {
    try {
      const { id } = req.params;

      // SECURITY: Validate and whitelist fields
      const validation = validateInventoryUpdate(req.body);

      if (!validation.valid) {
        return res.status(400).json({
          error: 'Validation failed',
          details: validation.errors
        });
      }

      const existingItem = await prisma.inventory.findUnique({
        where: { id: parseInt(id, 10) },
      });

      if (!existingItem) {
        return res.status(404).json({ error: 'Inventory item not found' });
      }

      const previousQuantity = existingItem.currentQuantity;

      // Update item with validated data only
      const updatedItem = await prisma.inventory.update({
        where: { id: parseInt(id, 10) },
        data: {
          ...validation.data,
          lastModifiedBy: req.user?.username || 'System',
          lastModifiedDate: new Date(),
        },
      });

      // Create audit entry if quantity changed
      if (validation.data!.currentQuantity !== undefined && validation.data!.currentQuantity !== previousQuantity) {
        await prisma.auditHistory.create({
          data: {
            itemId: updatedItem.id,
            itemNumber: updatedItem.itemNumber,
            itemDescription: updatedItem.description || updatedItem.hardwareDescription || 'No description',
            previousQuantity,
            newQuantity: validation.data!.currentQuantity,
            changedBy: req.user?.username || 'System',
            changeDate: new Date(),
          },
        });

        // Send low stock alert if needed
        if (
          validation.data!.currentQuantity < updatedItem.minimumThreshold &&
          previousQuantity >= updatedItem.minimumThreshold
        ) {
          try {
            await emailService.sendLowStockAlert(updatedItem, config.baseUrl);
          } catch (emailError) {
            console.error('Error sending low stock alert:', emailError);
          }
        }
      }

      res.json({
        ...updatedItem,
        needsReorder: updatedItem.currentQuantity < updatedItem.minimumThreshold,
      });
    } catch (error: any) {
      console.error('Error updating item:', error);
      res.status(500).json({ error: 'Failed to update inventory item' });
    }
  }

  /**
   * DELETE /api/inventory/:id
   * Delete inventory item (Admin only)
   */
  async deleteItem(req: Request, res: Response) {
    try {
      const { id } = req.params;

      const existingItem = await prisma.inventory.findUnique({
        where: { id: parseInt(id, 10) },
      });

      if (!existingItem) {
        return res.status(404).json({ error: 'Inventory item not found' });
      }

      // Create audit entry for deletion WITH item details
      // This record will persist after the item is deleted (itemId will be set to null)
      await prisma.auditHistory.create({
        data: {
          itemId: existingItem.id,
          itemNumber: existingItem.itemNumber,
          itemDescription: existingItem.description || existingItem.hardwareDescription || 'No description',
          previousQuantity: existingItem.currentQuantity,
          newQuantity: 0, // Set to 0 to indicate deletion
          changedBy: req.user?.username || 'System',
          changeDate: new Date(),
          serviceNowTicketUrl: 'ITEM DELETED',
        },
      });

      // Delete the inventory item
      await prisma.inventory.delete({
        where: { id: parseInt(id, 10) },
      });

      res.json({
        message: 'Inventory item deleted successfully',
        deletedItem: existingItem
      });
    } catch (error: any) {
      console.error('Error deleting item:', error);
      res.status(500).json({ error: 'Failed to delete inventory item' });
    }
  }
}

export const inventoryController = new InventoryController();
